src/dmd/s2ir.d

```diff
@@ -769,7 +769,8 @@ private extern (C++) class S2irVisitor : Visitor
                         Type t = ce.e1.type.toBasetype();
                         if (t.ty == Tdelegate)
                             t = t.nextOf();
-                        if (t.ty == Tfunction && retStyle(cast(TypeFunction)t) == RET.stack)
+                        if (t.ty == Tfunction &&
```

Comment:
You could then obtain do something like this at the call site...

```D
bool needsThis = fd && fd.needThis(();
TypeFunction tf = fd ? cast(TypeFunction)fd.type : t.ty;
Target.isReturnOnStack(tf, needsThis);
```

That's probably not correct (I haven't fully analyzed the code), but I hope it illustrates the idea.  Just get the `TypeFunction` and `needsThis` at the call site, and then make the call to `isReturnOnStack(tf, needsThis)`.

Reply:
That mostly brings us back to the initial version ;-) Agreed, passing `fd.needThis` instead of `fd` is the better API as it doesn't pass the same information twice which can cause ambiguities. It has to evaluate `needThis` eagerly, though. Considering all the stuff happening in isReturnOnStack, that's probably not a big deal.
I'll give that version another shot...